python/ray/data/dataset.py

```diff
@@ -228,12 +235,22 @@ def map(
             ray_remote_args: Additional resource requirements to request from
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
+        if isinstance(fn, CallableClass) and (
+            compute is None
+            or compute == "tasks"
+            or isinstance(compute, TaskPoolStrategy)
+        ):
+            raise ValueError(
```

Comment:
If this a must, why not just overwrite whatever from input args? Maybe just warn that it's overwritten in that case.

Reply:
The default is None, and this is raising exception on that default, that's weird...
The clearer semantics to me is we don't make this check, i.e. allow CallableClass to run with task pool, but log a warning on potential setup cost. This way the API is sane in terms of functionality, the difference only in performance.

Reply:
That sounds good to me, but I don't want to add that change in semantics as a driveby change in this PR, so let's surface this change in the default internally and open a follow-up PR.

Reply:
The issue is exactly that this part of code is  changing the API semantics:
- Before: map(CallableClass, TaskPoolStrategy) works
- After: it will raise exception

Reply:
No, as I said above, `map(CallableClass, TaskPoolStrategy)` current fails in master.

>Also to be clear, raising an error if a callable class and the default task pool strategy is provided is the current behavior, I just moved the error from within the map tasks to the API level for better error UX.

Reply:
>This worked on a previous release, so it must have changed lately.

We removed support for that since stateful tasks are an anti-pattern that hit weird reference counting edge cases and resulted in long-lived workers hogging clusters: https://github.com/ray-project/ray/pull/23708

> Then the action item here is to revise the documentation to reflect this new semantics.

The master API docs call out that callable classes are only supported for the actor compute model: https://docs.ray.io/en/master/data/package-ref.html#ray.data.Dataset.map_batches

And I add a dedicated [callable classes section](https://ray--24870.org.readthedocs.build/en/24870/data/transforming-datasets.html#callable-class-udfs) to the "Transforming Datasets" feature guide in the NumPy batch PR: https://github.com/ray-project/ray/pull/24870

So I think that docs are covered.

Reply:
Thanks for having them documented. For the docstring here, since this constraint is on "compute", it looks more natural to document it at "compute" arg than "fn" arg.

Reply:
Hmm but isn't the `fn` arg the best place to document it? Users are going to be looking at this UDF-first, where they have a UDF that involves some state initialization (e.g. cached loading of a model), so they'll look at the `fn` arg to see if it takes a class. If they see that a `CallableClass` is accepted without a note about only supporting the actor compute strategy, then they might miss a note in the `compute` arg docs warning that `compute="actors"` must be used in order to use a `CallableClass` UDF.

How about we document it in both places?

Reply:
I ended up adding a note about the constraint to both args in the map docstrings.